### PR TITLE
Nested folders: deduplicate selection of children

### PR DESCRIPTION
**Pull request by ashharrison90**
**What is this feature?**

- adds a new hook that retrieves the deduplicated selection state to be passed to the backend
- mocks out the backend api which will tell us affected children
- uses this mocked api in `MoveModal` and `DeleteModal` 
  - shows spinner when fetching, alert when error, etc.

**Why do we need this feature?**

- so we can delete the correct things in the backend

**Who is this feature for?**

- anyone using nested folders!

**Which issue(s) does this PR fix?**:

For https://github.com/grafana/grafana/issues/65679
For https://github.com/grafana/grafana/issues/65681

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/browse-dashboards/state/hooks.ts [hunk 1]
```diff
@@ -36,6 +44,10 @@ export function useSelectedItemsState() {
   return useSelector((wholeState: StoreState) => wholeState.browseDashboards.selectedItems);
 }
 
+export function useSelectedItemsForBackendState() {
```
Comment (ashharrison90) [inline public/app/features/browse-dashboards/state/hooks.ts:47]: naming is hard 😬 

wydt about renaming `useSelectedItemsState` and this to instead be `useCheckboxSelectionState` and `useApiSelectionState`? 🤔

Comment (joshhunt) [reply public/app/features/browse-dashboards/state/hooks.ts:47]: - `useCheckboxSelectionState` - sounds good 👍🏻 
 - `useAPISelectionState` - less certain on 'API', but I can't think of much better. `useActionSelectionState`? 🥴

File: public/app/features/browse-dashboards/state/hooks.ts [hunk 2]
```diff
@@ -83,3 +95,38 @@ function createFlatTree(
...
+    if (selectedItems.folder[parentFolderUID]) {
+      // Unselect any children
+      children?.forEach((child) => {
+        if (child.kind === 'dashboard') {
+          result.dashboard[child.uid] = false;
+        }
+        if (child.kind === 'folder') {
+          result.folder[child.uid] = false;
+        }
+      });
```
Comment (joshhunt) [inline public/app/features/browse-dashboards/state/hooks.ts:128]:
```suggestion
    // Unselect any children
    for (const child of children ?? []) {
      if (child.kind === 'dashboard') {
        result.dashboard[child.uid] = false;
      }
      if (child.kind === 'folder') {
        result.folder[child.uid] = false;
      }
    }
```

Annoying nit: I'm trying to prefer for-of/for-in over Array.forEach, especially when the forEach callback isn't a 'pure' function (it modifies `result` outside it's scope.

File: public/app/features/browse-dashboards/state/hooks.ts [hunk 3]
```diff
@@ -83,3 +95,38 @@ function createFlatTree(
...
 }
+
+/**
+ * Returns a DashboardTreeSelection but removes any selected folder's children.
+ * This is useful when making backend requests to move or delete items.
+ * In this case, we only need to move/delete the parent folder and it will cascade to the children.
+ * @param selectedItems Overall selection state
+ * @param childrenByParentUID Arrays of children keyed by their parent UID
+ */
+function getSelectedItemsForBackend(
```
Comment (joshhunt) [inline public/app/features/browse-dashboards/state/hooks.ts:106]: Took me a bit to properly grok how this works, but this is a really smart and neat solution. Well done 👍🏻

**Review by joshhunt — APPROVED** [review]
Looks good, and works well! 👍🏻

**Pull request merged**